### ISSUE TITLE
Implement MLJModelInterface.update (feature request)

### ISSUE THREAD

**Issue by ablaom** [IQVIA-ML/LightGBM.jl#36]
Implement`update` so that `fit!` only *adds* iterations when `model.n_iterations` increases -  that is, to suppress retraining "from scratch" in those cases. This is useful for generating out-of-sample learning curves (performance against `n_iterations`) and for controlling early stopping externally (using out-of-sample estimates of performance); see this [issue on iterative model control in MLJ ](https://github.com/alan-turing-institute/MLJ.jl/issues/139 )

You may also want to add a hyperparameter to your models to control whether or not changing `learning_rate` should trigger retraining from scratch. This will be useful for buying into more sophisticated control strategies, such as [cyclic learning rates](https://ieeexplore.ieee.org/abstract/document/7926641).  

To see how `fit!` behaves for models that implement `update` for an iteration parameter change, see the example [here](https://alan-turing-institute.github.io/MLJ.jl/dev/machines/). 

For example code I suggest looking at [this MLJFlux code](https://github.com/alan-turing-institute/MLJFlux.jl/blob/344ebe0e3e76178aa2b1588345cbd41bdebe9d33/src/regressor.jl#L93) which has an  `optimiser_changes_trigger_retraining` control hyperparameter (the optimiser stores the learning rate and related parameters). The MLJFlux code also includes good tests for checking the update logic works. Other examples are EvoTrees.jl (pure julia tree-boosting) and the EnsembleModel wrap (at MLJ/src/ensembles.jl).

**Comment by yalwan-iqvia** [comment]
@ablaom Thanks for raising this request. 

If you can think of other aspects of MLJ integration which might be missing, please do raise them as issues. As users might start using the package they might well ask for things which are from the python package, and where possible I'd like to be able to point them to MLJ integrations rather than implement similar features from scratch.

**Comment by ablaom** [comment]
As an example of where this would help out is the learning curves for n_rounds you put in the tutorial. They currently run O(N^2) where N is the upper limit, but will be O(N) if `update` is implemented. Of course on Boston, you hardly notice the difference.

**Comment by yalwan-iqvia** [comment]
I'm thinking about what I need to do to implement this, and I believe it would be sufficient to premit it for increases on `n_iterations` only -- the reason being that while cyclic learning rates are a useful notion, gradient boosting solves a convex problem and so I believe cyclic learning rates are unnecessary.

**Comment by ablaom** [comment]
Well,  cyclic learning rates are just one control strategy one might want to implement. However, adapting the learning rate during training is "bells and whistles". Be very happy to see an update to catch `n_iterations` increases, as this is really essential.

**Comment by yalwan-iqvia** [comment]
Just one question: How come it is `update` rather than `update!` ?

**Comment by yalwan-iqvia** [comment]
Would you also be able to supply a usage snippet?

**Comment by ablaom** [comment]
> Just one question: How come it is update rather than update! ?

It's not anticipated that you would need to mutate arguments. This low-level model interface is "functional" style: You suck in a `fitresult` and `cache` and spit out new ones. See https://alan-turing-institute.github.io/MLJ.jl/dev/adding_models_for_general_use/#Iterative-models-and-the-update!-method-1  Do you think you might need to mutate something?

> Would you also be able to supply a usage snippet?

Maybe you misunderstand. This method is not for the user; rather it is something that will enhance the behavior of `fit!(::Machine)`, as in this example here, after "However, for this iterative model, " : https://alan-turing-institute.github.io/MLJ.jl/dev/machines/#Machines-1 . A simplified version of the logic executed in `fit!` (which decides whether to call `fit` or `update`)  is here: https://alan-turing-institute.github.io/MLJ.jl/dev/internals/#fit!-and-predict/transform-1

The EvoTrees.jl implementation is here:
https://github.com/Evovest/EvoTrees.jl/blob/4235aa2548ce93a60ebe6df18f9909fa856ee71a/src/MLJ.jl#L24 (But you can avoid the length `okay_to_continue` method by using `MLJModelInterface.is_same_except`; query the doc-string). This is very close to your case - it's another tree-booster. See also the Flux example quoted earlier. 

Does that help?

**Comment by yaxxie** [comment]
I think so, mostly I am concerned with ensuring I implement it properly and doing what is needed of it, rather than something way off-piste.

As for mutating things, no, I had a slightly dumb moment (I had observed fit was functional prior to this and then apparently forgotten), no further remarks there.

**Comment by yalwan-iqvia** [comment]
Ok thats good after spending a few moments on EvoTrees one it looks like they do the same as what I say, if number of rounds increases only, we can update else we must fit.

**Comment by yalwan-iqvia** [comment]
Please take a look at this PR: https://github.com/IQVIA-ML/LightGBM.jl/pull/51

Its in draft at the moment while I figure out how much tidying up I want to do as a consequence
All your remarks are appreciated
I would also appreciate if you can provide some snippet code I can use to exercise the `update` path in an integration style test
I didn't write any new tests for `update` method yet

**N.B.** I'm not available next week, just FYI.

**Comment by ablaom** [comment]
Yeah, testing this is a bit finicky. One way we do this - and it may not be the best - is with `@test_logs`. The idea is that you set the verbosity to a level where each new iteration round makes some `@info ` call (I don't know if you have that) which allows you to count how many iterations actually get performed. You can have a look at the Flux tests here, for example: https://github.com/alan-turing-institute/MLJFlux.jl/blob/3ffd02b4ccf17a2f840ec5f8155c1b6ee05b016a/test/test_utils.jl#L25

Or you can test at the level of machines and do something similar:
https://github.com/alan-turing-institute/MLJ.jl/blob/4062aab42f29b1c5d0410a68c12806ad8774ca42/test/ensembles.jl#L219

**Comment by yalwan-iqvia** [comment]
https://github.com/IQVIA-ML/LightGBM.jl/pull/51

I un-drafted the PR, if you can take a look it would be appreciated.

**Comment by yalwan-iqvia** [comment]
PR is merged (plus mini fixing PR)

Closing this.
